libp2p/module.go

```diff
@@ -0,0 +1,367 @@
+	// (i.e. multiple, unidentified receivers)
+	pubsub *pubsub.PubSub
+	// topic similar to pubsub but received messages are filtered by a "topic" string.
+	// Published messages are also given the respective topic before broadcast.
+	topic *pubsub.Topic
+	// subscription provides an interface to continuously read messages from.
+	subscription *pubsub.Subscription
+	network      typesP2P.Network
+}
+
+var (
+	// TECHDEBT: configure timeouts. Consider security exposure vs. real-world conditions).
+	// TECHDEBT: parameterize and expose via config.
+	// readStreamTimeout is the duration to wait for a read operation on a
+	// stream to complete, after which the stream is closed ("timed out").
+	readStreamTimeoutDuration = time.Second * 10
+)
+
+func Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, error) {
+	return new(libp2pModule).Create(bus, options...)
+}
+
+func (mod *libp2pModule) GetModuleName() string {
+	return modules.P2PModuleName
+}
+
+func (mod *libp2pModule) Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, error) {
+	logger.Global.Print("Creating libp2p-backed network module")
+
+	*mod = libp2pModule{
+		cfg: bus.GetRuntimeMgr().GetConfig().P2P,
+	}
+
+	// MUST call before referencing mod.bus to ensure != nil.
+	bus.RegisterModule(mod)
+
+	for _, option := range options {
+		option(mod)
+	}
+
+	// TECHDEBT: investigate any unnecessary
+	// key exposure / duplication in memory
+	privateKey, err := crypto.NewLibP2PPrivateKey(mod.cfg.PrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("loading private key: %w", err)
+	}
+
+	mod.identity = libp2p.Identity(privateKey)
+
+	// INCOMPLETE: support RainTree network
+	if mod.cfg.UseRainTree {
+		return nil, fmt.Errorf("%s", "raintree is not yet compatible with libp2p")
+	}
+
+	switch mod.cfg.ConnectionType {
+	case types.ConnectionType_TCPConnection:
+		addr, err := mod.getMultiaddr()
+		if err != nil {
+			return nil, fmt.Errorf("parsing multiaddr from config: %w", err)
+		}
+		mod.listenAddrs = libp2p.ListenAddrs(addr)
+	case types.ConnectionType_EmptyConnection:
+		mod.listenAddrs = libp2p.NoListenAddrs
+	default:
+		return nil, fmt.Errorf(
+			// DISCUSS: rename to "transport protocol" instead.
+			"unsupported connection type: %s: %w",
+			mod.cfg.ConnectionType,
+			err,
+		)
+	}
+
+	return mod, nil
+}
+
+func (mod *libp2pModule) Start() error {
+	mod.logger = logger.Global.CreateLoggerForModule(mod.GetModuleName())
+
+	// IMPROVE: receive context in interface methods.
+	ctx := context.Background()
+
+	// TECHDEBT: metrics integration.
+	var err error
+	opts := []libp2p.Option{
+		mod.identity,
+		// INCOMPLETE(#544): add transport security!
+	}
+
+	// Disable unused libp2p relay and ping services in client debug mode.
+	// (see: https://pkg.go.dev/github.com/libp2p/go-libp2p#DisableRelay
+	// and https://pkg.go.dev/github.com/libp2p/go-libp2p#Ping)
+	if mod.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode {
```

Comment:
Not a strong opinion but: `mod.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode` is used here and below, perhaps we can wrap it into a bool `isClientDebugMode` or something.